File: customer_info/customer_info/page/payments_received/payments_received.js
frappe.pages['payments-received'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Payments Received',
		single_column: true
	});
	wrapper.payments_received = new payments_received(wrapper)
	frappe.breadcrumbs.add("Customer Info");
}

payments_received = Class.extend({
	init: function(wrapper) {
		var me = this;
		this.wrapper_page = wrapper.page;
		this.page = $(wrapper).find('.layout-main-section-wrapper');
		this.wrapper = $(wrapper).find('.page-content');
		export_btn: wrapper.page.set_secondary_action(__("Export"),
		function() { 
					me.get_data_export();
				}, "icon-refresh");
		this.set_fields();
		//this.render_payments_details();
	},
	set_fields: function() {
		var me = this;
		html = "<div>\
				<div class='col-xs-2 customer'></div>\
				<div class='col-xs-2 from_date'></div>\
  				<div class='col-xs-2 to_date'></div>\
  				<div class='col-xs-2 agreement'></div>\
  				<div class='col-xs-2 data_limit'></div>\
  				<div class='col-xs-2 payment_type_filter'></div>\
  				<div class='col-xs-2' ></div>\
  				</div>\
				<table id='tableSearchResults' class='table table-hover  table-striped table-condensed' style='font-size:12px;margin-bottom: 0px;'>\
			     	<thead>\
			            <tr>\
			                <th width='7%'>Payment Date</th>\
			                <th width='7%'>Customer Name</th>\
			                <th width='7%'>Payment Type</th>\
			                <th width='7%'>Payment Amount</th>\
			                <th width='6%'>Late Fees Amount</th>\
			                <th width='5%'>Receivables Amount</th>\
			                <th width='5%'>Discount Amount</th>\
			                <th width='5%'>Campaign Discount Amount</th>\
			                <th width='5%'>Bonus Amount</th>\
			                <th width='13%'>Total calculated payment amount</th>\
			                <th width='11%'>Bank Transfer Amount</th>\
			                <th width='5%'>Receivables Collected</th>\
			                <th width='5%'>Cash Amount</th>\
			                <th width='3%'>Bank Card Amount</th>\
			                <th width='4%'>Associate</th>\
			                <th width='3%'>Refund Payment</th>\
			                <th width='2%'></th>\
			            </tr>\
			        </thead></table>"
		me.page.html(html)
		me.customer_link = frappe.ui.form.make_control({
			parent: me.page.find(".customer"),
			df: {
			fieldtype: "Link",
			options: "Customer",
			fieldname: "customer",
			label:"Customer",
			placeholder: "Select Customer"
			},
			render_input: true
		});
		me.customer_link.refresh();
		me.from_date = frappe.ui.form.make_control({
			parent: me.page.find(".from_date"),
			df: {
				fieldtype: "Date",
				fieldname: "from_date",
				label:"From Date",
				placeholder: "From Date"
			},
			render_input: true
		});
		me.from_date.refresh();
		me.to_date = frappe.ui.form.make_control({
			parent: me.page.find(".to_date"),
			df: {
				fieldtype: "Date",
				fieldname: "to_date",
				label:"To Date",
				placeholder: "To Date"
			},
			render_input: true
		});
		me.to_date.refresh();
		me.agreement = frappe.ui.form.make_control({
			parent: me.page.find(".agreement"),
			df: {
				fieldtype: "Link",
				fieldname: "agreement",
				label:"Customer Agreement",
				placeholder: "Customer Agreement",
				options:"Customer Agreement"
			},
			render_input: true
		});
		me.agreement.refresh();

		me.data_limit = frappe.ui.form.make_control({
			parent: me.page.find(".data_limit"),
			df: {
				fieldtype: "Select",
				fieldname: "data_limit",
				label:"Data Limit",
				placeholder: "Data Limit",
				options:["","0-100","0-200","0-300","0-400","0-500","All"]
			},
			render_input: true
		});

		me.payment_type_filter = frappe.ui.form.make_control({
		parent: me.page.find(".payment_type_filter"),
		df: {
				fieldtype: "Select",
				fieldname: "payment_type_filter",
				label:"Payment Type",
				placeholder: "payment_type",
				options:["","Rental Payment","Modification Of Receivables","90d SAC"]
			},
			render_input: true
		});
		me.payment_type_filter.refresh();
		
		this.render_payments_details();	

		me.customer_link.$input.on("change", function(){
			var old_me = me;
			old_me.render_payments_details()
		});

		me.from_date.$input.on("change", function(){
			var old_me = me;
			old_me.render_payments_details()
		});

		me.to_date.$input.on("change", function(){
			var old_me = me;
			old_me.render_payments_details()
		});

		me.agreement.$input.on("change", function(){
			var old_me = me;
			old_me.render_payments_details()
		});

		me.data_limit.$input.on("change", function(){
			var old_me = me;
			old_me.render_payments_details()
		});

		me.payment_type_filter.$input.on("change", function(){
			var old_me = me;
			old_me.render_payments_details()
		});
	},
	render_payments_details: function() {
		var me = this;
		//var _agreement_name =  window.location.href.split("=")[1] ? window.location.href.split("=")[1]:""
		//me.agreement.$input.val(_agreement_name)
		var me = this;
		console.log("------------ggg-----",me.payment_type_filter.$input.val())
		this.data = ""
		frappe.call({
			method: "customer_info.customer_info.page.payments_received.payments_received.get_payments_details",
			args: {
				"customer": me.customer_link.$input.val(),
				"from_date": me.from_date.$input.val(),
				"to_date":me.to_date.$input.val(),
				"agreement":me.agreement.$input.val(),// ? me.agreement.$input.val():_agreement_name,
				"data_limit":me.data_limit.$input.val(),
				"pmt_type" :me.payment_type_filter.$input.val()	
			},
			freeze: true,
			freeze_message: __("Please Wait..."),
			callback: function(r) {
        	   	me.page.find(".data").empty();
				$.each(r.message["data"], function(i, d) {
					/*if(d["payoff_cond"]){
						me.payoff_cond = d["payoff_cond"]
					}
					if(d["payment_type"]){
						me.payment_type = d["payment_type"]
					}*/
					me.payoff_cond = d["payoff_cond"] ? d["payoff_cond"] : ""
					me.payment_type = d["payment_type"] ? d["payment_type"]: ""
					me.payments_ids = d["payments_ids"]
					me.late_fees_updated = d["late_fees_updated"]
					me.updated_late_fees = d["late_fees"]
					d["payments_ids"] = d["payments_ids"] ? me.update_dict_by_payment_ids():""
        	   	});
        	   	me.data = r.message["data"];
        	   	html = frappe.render_template("payments_received",{"data": r.message["data"],"total":r.message["total"]})
				me.page.append(html)
				me.refund();
			}
		})
	},
	update_dict_by_payment_ids:function(){
		var me = this;
		var flt_precision = frappe.defaults.get_default("float_precision")
		var dict_of_payments_ids = []
		var __dict_of_payments_ids = []
		if(me.payment_type != "Modification Of Receivables"){
			//console.log(me.payments_ids,"payments_ids12122121",String(me.payments_ids.slice(1,-1)).split(","))
			var formatted_list_of_payment_ids = String(me.payments_ids.slice(1,-1)).split(",")//JSON.parse("[" + me.payments_ids.slice(0,-1) + '"' + "]")[0].split(",")
			if(me.payoff_cond != "Rental Payment"){
				late_fees = ""
				rental_payment = 0
				total = ""
				payment_id = ""
				payments_ids = []
				$.each(formatted_list_of_payment_ids, function(i, d) {
					payment_id = d.split("/")[0].split("-P")[0],
					payments_ids.push(d.split("/")[0]),
					late_fees = "-",
					rental_payment = parseFloat(flt(d.split("/")[2]).toFixed(2)),
					total = "-"
		   		});
		   		__dict_of_payments_ids.push({
		   			"payments_id":payment_id+"-"+me.payoff_cond,
					"payment_id_list": JSON.stringify(payments_ids.toString()),
					"due_date":"-",
					"rental_payment":rental_payment.toFixed(2),
					"late_fees":late_fees,
					"total": total
		   		})
				return __dict_of_payments_ids
			}
			else{
				if(me.late_fees_updated == "Yes"){
					//console.log("inside late_fees_updated")
					/*oldest_date = {"payments_id":"","due_date":""}
					var due_date_list = []
					$.each(formatted_list_of_payment_ids, function(i, d) {
						console.log(d.split("/")[0],i,"$$$$$$$$%%%%%%%%%%%%%%")
						due_date_list.push(d.split("/")[1])
						function dmyOrdA(a,b){ return myDate(a) - myDate(b);}
						function myDate(s){var a=s.split(/-|\//); return new Date(a[0],a[1]-1,a[2]);}
						due_date_list.sort(dmyOrdA);
						oldest_date["due_date"] = due_date_list[0]
						if(oldest_date["due_date"] == d.split("/")[1]){
							oldest_date["payments_id"] = d.split("/")[0]
						}
					})*/
				   	$.each(formatted_list_of_payment_ids, function(i, d) {
				   		//if (d.split("/")[0] == oldest_date["payments_id"]){
				   		if (i == 0){
				   			dict_of_payments_ids.push({
				   				"late_fees": me.updated_late_fees,
				   				"payments_id":d.split("/")[0],
								"due_date":d.split("/")[1],
								"rental_payment":d.split("/")[2],
								"total": parseFloat(flt(me.updated_late_fees) + flt(d.split("/")[2])).toFixed(2) 
				   			})
				   		}	
				   		else{
				   			dict_of_payments_ids.push({	
				   				"late_fees":0,
				   				"payments_id":d.split("/")[0],
								"due_date":d.split("/")[1],
								"rental_payment":d.split("/")[2],
								"total": parseFloat(0 + flt(d.split("/")[2])).toFixed(2) 
				   			})
				   		}
			   		});
				}
				else{
					//console.log("in else inside late_fees_updated")
					$.each(formatted_list_of_payment_ids, function(i, d) {
				   		dict_of_payments_ids.push({
				   			"payments_id":d.split("/")[0],
							"due_date":d.split("/")[1],
							"rental_payment":d.split("/")[2],
							"late_fees":parseFloat(me.get_late_fees(d.split("/")[0].split("-P")[0],d.split("/")[1],d.split("/")[3],d.split("/")[2])).toFixed(2),
							"total": parseFloat(flt(me.get_late_fees(d.split("/")[0].split("-P")[0],d.split("/")[1],d.split("/")[3],d.split("/")[2])) + flt(d.split("/")[2])).toFixed(2)
							//parseFloat(me.get_late_fees(d.split("/")[0].split("-P")[0],d.split("/")[1],d.split("/")[3],d.split("/")[2]) + flt(d.split("/")[2])).toFixed(2) 
				   		})
			   		});
				}
				return dict_of_payments_ids
			}
		}	
	},
	get_late_fees : function(agreement_name,date1,date2,rental_payment){
		//console.log(agreement_name,"agreement_name",date1,date2)
		var date_diff = frappe.datetime.get_diff(date2,date1)
		var late_fees_amount = 0
		frappe.call({
			async:false,
            method: "frappe.client.get_value",
            args: {
                doctype: "Customer Agreement",
                fieldname: "late_fees_rate",
                filters: { name: agreement_name },
            },
            callback: function(res){
            //	console.log("********************8888")
                if (res && res.message){
                //	console.log("rerer",res.message,date_diff)
					if(flt(date_diff) > 3){
				//		console.log("insdie 333333333")
						late_fees_amount = (flt(date_diff) - 3) * rental_payment * (flt(res.message.late_fees_rate)/100)
					}
                }
            }
        });
		if(flt(date_diff) > 3 && late_fees_amount){
			return late_fees_amount.toFixed(2)
		}
		else{
			return 0
		}
	},
	refund:function(){
		var me = this;		
		$(me.page).find(".refund").click(function() {
			me.ph_name = $(this).attr("ph-name")
			me.show_dialog();
		});
	},
	show_dialog:function(){
		var me = this;
		this.dialog = new frappe.ui.Dialog({
            		title: "Refund Process",
                	fields: [
                   		{"fieldtype": "Select" , "fieldname": "refund" , "label": "Do You Want To Refund","options":["No","Yes"],"default":"No"},
                   		{"fieldtype": "Button" , "fieldname": "refund_payment" , "label": "Refund"}
                   	],
	       		});
       	this.fd = this.dialog.fields_dict;
       	this.dialog.$wrapper.find('.modal-dialog').css("width", "350px");
       	this.dialog.$wrapper.find('.modal-dialog').css("height", "300px");
       	this.dialog.$wrapper.find('.hidden-xs').css("margin-left","-2px");
       	this.dialog.show();
       	me.click_on_refund_inside_dialog();
	},
	click_on_refund_inside_dialog:function(){
		var me = this;
		me.fd.refund.$wrapper.find('[data-fieldtype="Select"]').attr("dir","rtl");
		me.fd.refund_payment.$wrapper.find('[data-fieldname="refund_payment"]').css({"float":"right"});
		me.fd.refund_payment.$input.click(function() {
			me.make_refund_payment()
		})	
	},
	make_refund_payment:function(){
		var me = this;
		if(me.fd.refund.$input.val() == "Yes"){
			var id_list = []
			len = $('.'+String(me.ph_name)).length
			if($('.'+String(me.ph_name)).attr("payment-ids")){
				id_list = $('.'+String(me.ph_name)).attr("payment-ids").split(",")
			}
			else{
				for(i=0;i<len;i++){
					id_list.push($($('.'+String(me.ph_name))[i]).text())
				}
			}
			//if(id_list.length>0){
			//}
			frappe.call({
		        method: "customer_info.customer_info.page.payments_received.payments_received.make_refund_payment",
	           	args: {
	           		"payments_ids":id_list,
	            	"ph_name":me.ph_name ? me.ph_name :""
	            },
		       	callback: function(r){
		       		$('tr#'+String(me.ph_name)).hide()
		       		me.dialog.hide();
		       		me.render_payments_details();
		    	}
	    	});
		}
		else{
			me.dialog.hide();
		}
	},
	get_data_export:function(){
		var me = this;
		/*window.location.href = repl(frappe.request.url +
		'?cmd=%(cmd)s&data=%(data)s', {
			cmd: "customer_info.customer_info.page.payments_received.payments_received.create_csv",
			data:JSON.stringify(me["data"])
		});*/
		window.location.href = repl(frappe.request.url +
		'?cmd=%(cmd)s&from_date=%(from_date)s&to_date=%(to_date)s&customer=%(customer)s&agreement=%(agreement)s&data_limit=%(data_limit)s', {
			cmd: "customer_info.customer_info.page.payments_received.payments_received.create_csv",
			from_date: me.from_date.$input.val(),
			to_date: me.to_date.$input.val(),
			customer: me.customer_link.$input.val(),
			agreement: me.agreement.$input.val(),
			data_limit: me.data_limit.$input.val(),
		});
	}
});

File: customer_info/customer_info/page/payments_received/payments_received.py
from __future__ import unicode_literals
import frappe
import json
from frappe.utils import flt, cstr, cint, date_diff
from frappe.utils.csvutils import UnicodeWriter
import pdfkit
from customer_info.customer_info.doctype.payments_management.payments_management import set_values_in_agreement_on_submit,set_values_in_agreement_temporary 

@frappe.whitelist()
def get_payments_details(customer,from_date,to_date,agreement,data_limit):

	if customer and from_date and to_date:
		cond = "where customer = '{0}' and (payment_date BETWEEN '{1}' AND '{2}') and refund = 'No' ".format(customer,from_date,to_date)

	elif customer and from_date:
		cond = "where customer = '{0}' and payment_date >= '{1}' and refund = 'No' ".format(customer,from_date)

	elif customer and to_date:
		cond = "where customer = '{0}' and payment_date < '{1}' and refund = 'No' ".format(customer,to_date)

	elif from_date and to_date:
		cond = "where (payment_date BETWEEN '{0}' AND '{1}')  and refund = 'No' ".format(from_date,to_date)

	elif customer:
		cond = "where customer = '{0}'  and refund = 'No' ".format(customer)

	elif from_date:
		cond = "where payment_date >= '{0}' and refund = 'No' ".format(from_date)

	elif to_date:
		cond = "where payment_date <= '{0}' and refund = 'No' ".format(to_date)

	else:
		cond = " where refund = 'No' "

	_cond = "limit 50" if not agreement	and not from_date and not to_date and not customer else " "
	if data_limit and not agreement and not from_date and not to_date and not customer:
		if data_limit != "All":
			_cond = "limit {0}".format(data_limit.split('-')[1])
		else:
			_cond = " "


	data = frappe.db.sql("""select payment_date,customer,payoff_cond,
								format(rental_payment+campaign_discount,2) as rental_payments,
								format(1*late_fees,2) as late_fees,receivables,
								CASE WHEN payoff_cond = "Rental Payment" 
								THEN format(rental_payment+late_fees-receivables-bonus-discount,2) ELSE format(total_payment_received,2) END AS total_payment_received,
	    						format(bank_transfer,2) as bank_transfer,format(cash,2) as cash,receivables_collected,format(bank_card,2) as bank_card,
								balance,format(discount, 2) as discount,format(campaign_discount, 2) as campaign_discount,format(bonus,2) as bonus,concat(name,'') as refund,payments_ids,
								late_fees_updated,payment_type,merchandise_status,
								case when special_associate = "Automatic" then special_associate else owner end as associate
								from `tabPayments History` {0}
								order by payment_date desc {1} """.format(cond,_cond),as_dict=1)


	filter_data = []
	filter_payments_history = []
	for row in data:
		if agreement and row['payments_ids'] and agreement in [i.split('-P')[0] for i in str(row['payments_ids'])[1::].split(',')[0:-1]]:
			filter_data.append(row)
			filter_payments_history.append(row['refund'])

	filter_payments_history = [name.encode('utf-8') for name in filter_payments_history]

	if agreement:
		if filter_payments_history:
			if len(filter_payments_history) == 1:
				cond += " and name = '{0}' ".format(filter_payments_history[0])
			elif len(filter_payments_history) > 1:
				cond += " and name in {0}".format(tuple(filter_payments_history))

		data = filter_data

	total = frappe.db.sql("""select "payment_date" as payment_date,"customer" as customer,"payoff_cond" as payoff_cond,
						format(sum(rental_payment),2) as rental_payment,
						format(sum(1*late_fees),2) as late_fees,format(sum(receivables),2) as receivables,"total_payment_received" as total_payment_received,
						format(sum(bank_transfer),2) as bank_transfer,format(sum(cash),2) as cash ,sum(receivables_collected) as receivables_collected,format(sum(bank_card),2) as bank_card,
						format(sum(balance),2) as balance,format(sum(discount),2) as discount,format(sum(campaign_discount),2) as campaign_discount, format(sum(bonus),2) as bonus
						from `tabPayments History` {0}""".format(cond,_cond),as_dict=1)
	total_payment_received = []

	for row in data:
		if row.get("associate") == "Administrator":
			row['associate'] = frappe.db.get_value("User",{'first_name':row['associate']},"first_name")
		elif row.get("associate") == "Automatic":
			pass
		else:		
			row['associate'] = frappe.db.get_value("User",{'email':row['associate']},"first_name")
		total_payment_received.append(row['total_payment_received'].replace(",",""))

	total[0]["payment_date"] = "Total"
	total[0]["customer"] = "-"
	total[0]["payoff_cond"] = "-"
	total[0]['total_payment_received'] = "{0:.2f}".format(sum(map(float,total_payment_received)))

	return {"data":data,"total":total}

@frappe.whitelist()
def create_csv(customer,from_date,to_date,agreement,data_limit):
	data_list = get_payments_details(customer,from_date,to_date,agreement,data_limit)['data']
	for row in data_list:
			row["payments_ids"] = update_dict_by_payment_ids(row) if row.get("payments_ids") else ""
	data = data_list
	w = UnicodeWriter()
	w = add_header(w)
	w = add_data(w, data)
	# write out response as a type csv
	frappe.response['result'] = cstr(w.getvalue())
	frappe.response['type'] = 'csv'
	frappe.response['doctype'] = "Payment Received Report"


def update_dict_by_payment_ids(row):
	dict_of_payments_ids = []
	__dict_of_payments_ids = []
	if row.payment_type != "Modification Of Receivables":
		formatted_list_of_payment_ids = str(row.payments_ids[1:-1]).split(",")
		if row.payoff_cond != "Rental Payment":
			late_fees = ""
			rental_payment = 0
			total = ""
			payment_id = ""
			payments_ids = []
			for d in formatted_list_of_payment_ids:
				payment_id = d.split("/")[0].split("-P")[0],
				payments_ids.append(d.split("/")[0]),
				late_fees = "-",
				rental_payment = float(d.split("/")[2]),
				total = "-"
	   		__dict_of_payments_ids.append({
	   			"payments_id":str(payment_id)+"-"+row.payoff_cond,
				"payment_id_list": str(payments_ids),
				"due_date":"-",
				"rental_payment":rental_payment,
				"late_fees":late_fees,
				"total": total
	   		})
			return __dict_of_payments_ids

		else:
			if row.late_fees_updated == "Yes":
				for index,data in enumerate(formatted_list_of_payment_ids):
					if index == 0:
			   			dict_of_payments_ids.append({
			   				"late_fees": row.updated_late_fees,
			   				"payments_id":data.split("/")[0],
							"due_date":data.split("/")[1],
							"rental_payment":data.split("/")[2],
							"total": float(float(row.updated_late_fees) if row.updated_late_fees else 0 + float(data.split("/")[2])) 
			   			})	
			   		else:
			   			dict_of_payments_ids.append({	
			   				"late_fees":0,
			   				"payments_id":data.split("/")[0],
							"due_date":data.split("/")[1],
							"rental_payment":data.split("/")[2],
							"total": float(0 + float(data.split("/")[2])) 
			   			})
			else:
		   		for d in formatted_list_of_payment_ids:
		   			dict_of_payments_ids.append({
			   			"payments_id":d.split("/")[0],
						"due_date":d.split("/")[1] if d.split("/")[1] else "",
						"rental_payment":d.split("/")[2],
						"late_fees":float(get_late_fees(d.split("/")[0].split("-P")[0],d.split("/")[1],d.split("/")[3],d.split("/")[2])),
						"total": float(float(get_late_fees(d.split("/")[0].split("-P")[0],d.split("/")[1],d.split("/")[3],d.split("/")[2])) + float(d.split("/")[2]))
			   		})
			return dict_of_payments_ids


def get_late_fees(agreement_name,date1,date2,rental_payment):
	date_diffirence = date_diff(date2,date1)
	late_fees_amount = 0
	agreement_doc = frappe.get_doc("Customer Agreement",agreement_name)
	if flt(date_diffirence) > 3:
		late_fees_amount = (float(date_diffirence) - 3) * float(rental_payment) * (float(agreement_doc.late_fees_rate)/100)
	if float(date_diffirence) > 3 and late_fees_amount:
		return late_fees_amount
	else:
		return 0

def add_header(w):
	w.writerow(["Payment Received Report"])
	return w

def add_data(w,data):
	if len(data) > 0:
		w.writerow('\n')
		w.writerow(['Payment Received'])
		w.writerow(['', 'Payment Date', 'Customer Name', 'Payment Type', 'Payment Amount',\
					 'Late Fees Amount', 'Receivables Amount','Discount Amount',\
					 'Campaign Discount Amount', 'Bonus Amount', 'Total Calculated Payment Amount',\
					  'Bank Transfer Amount', 'Receivables Collected', 'Cash Amount', 'Bank Card Amount'])
		for i in data:
			row = ['',i['payment_date'], i['customer'], i['payment_type'],i['rental_payments']\
				,i['late_fees'],i['receivables'],i['discount'],\
				i['campaign_discount'],i['bonus'],i['total_payment_received'],i['bank_transfer']\
				,i['receivables_collected'],i['cash'],i['bank_card']]
			w.writerow(row)	
			w.writerow(['','','Payment id','Due Date','Rental Payment','Late Fees','Total'])
			for j in i['payments_ids']:
				row = ['','', j['payments_id'],j['due_date'],j['rental_payment'],j['late_fees'],j['total']]
				w.writerow(row)
	return w

# @frappe.whitelist()
# def create_csv(data):
# 	w = UnicodeWriter()
# 	w = add_header(w)
# 	w = add_data(w, data)
# 	# write out response as a type csv
# 	frappe.response['result'] = cstr(w.getvalue())
# 	frappe.response['type'] = 'csv'
# 	frappe.response['doctype'] = "Payment Received Report"

# def add_header(w):
# 	w.writerow(["Payment Received Report"])
# 	return w

# def add_data(w,data):
# 	data = json.loads(data)
# 	if len(data) > 0:
# 		w.writerow('\n')
# 		w.writerow(['Payment Received'])
# 		w.writerow(['', 'Payment Date','Customer', 'Rental Payment','Late Fees','Receivables','Total Rental Payment','Bank Transfer','Cash','Bank Card','Balance','Discount','Bonus'])
# 		for i in data:
# 			row = ['',i['payment_date'], i['customer'], i['rental_payment'],i['late_fees'],i['receivables'],i['total_payment_received'],i['bank_transfer'],i['cash'],i['bank_card'],i['balance'],i['discount'],i['bonus']]
# 			w.writerow(row)	
# 			w.writerow(['','Payment id','Due Date','Rental Payment','Late Fees','Total','','Payment id','Due Date','Rental Payment','Late Fees','Total'])
# 			for j in i['payments_ids']:
# 				row = ['', j['payments_id'],j['due_date'],j['rental_payment'],j['late_fees'],j['total'],'', j['payments_id'],j['due_date'],j['rental_payment'],j['late_fees'],j['total']]
# 				w.writerow(row)
# 	return w
	 
@frappe.whitelist()
def make_refund_payment(payments_ids,ph_name):
	payments_ids = json.loads(payments_ids)
	payment_history = frappe.get_doc("Payments History",ph_name)
	customer = frappe.get_doc("Customer",payment_history.customer)
	payments_id_list = []
	agreement_list = []
	merchandise_status_list= []
	campaign_discount_of_agreements_list = []
	if len(payments_ids) > 0:
		for i in payments_ids:
			frappe.db.sql("""update `tabPayments Record` set check_box = 0,pre_select_uncheck = 0,
								payment_date = "",check_box_of_submit = 0,payment_history = "",pmt="",associate="",
								total_transaction_amount = 0 
								where check_box_of_submit = 1 
								and payment_id = '{0}' """.format(i))
			payments_id_list.append(i)
			agreement_list.append(i.split("-P")[0])
		agreement_list =  list(set(agreement_list))
		if agreement_list:
			agreement_list = [x.encode('UTF8') for x in agreement_list if x]
		flag = "Make Refund"
		agreement_list.sort()

		merchandise_status = payment_history.merchandise_status
		if merchandise_status and payment_history.payment_type == "Normal Payment":
			merchandise_status_list = [x.encode('UTF8') for x in merchandise_status.split(",")[0:-1] if x]	
			merchandise_status_list.sort()


		campaign_discount_of_agreements = payment_history.campaign_discount_of_agreements
		if campaign_discount_of_agreements and payment_history.payment_type == "Normal Payment":
			
			campaign_discount_of_agreements_list = [x.encode('UTF8') for x in campaign_discount_of_agreements.split(",")[0:-1] if x]	
			campaign_discount_of_agreements_list.sort()
			campaign_discount_of_agreements_dict = { agreement.split("/")[0]:[agreement.split("/")[1],agreement.split("/")[2]] for agreement in campaign_discount_of_agreements_list}

		refund_bonus = []	
		for i,agreement in enumerate(agreement_list):
			customer_agreement = frappe.get_doc("Customer Agreement",agreement)
			set_values_in_agreement_on_submit(customer_agreement)
			item_doc = frappe.get_doc("Item",customer_agreement.product)
			if payment_history.payment_type == "Payoff Payment":
				payment_history.payoff_cond = ""
				item_doc.sold_date = item_doc.old_sold_date
				item_doc.old_sold_date = item_doc.old_sold_date
				item_doc.save(ignore_permissions=True)
				customer_agreement.agreement_status = "Open"
				customer_agreement.old_merchandise_status = customer_agreement.merchandise_status
				customer_agreement.merchandise_status = payment_history.merchandise_status
				customer_agreement.agreement_closing_suspending_reason = ""
				customer_agreement.save(ignore_permissions=True)

			if payment_history.payment_type == "Normal Payment" and agreement == merchandise_status_list[i].split("/")[0]:
				item_doc.sold_date = item_doc.old_sold_date
				item_doc.old_sold_date = item_doc.old_sold_date
				item_doc.save(ignore_permissions=True)
				customer_agreement.agreement_status = "Open"
				customer_agreement.old_merchandise_status = customer_agreement.merchandise_status
				customer_agreement.merchandise_status = merchandise_status_list[i].split("/")[1]
				customer_agreement.agreement_closing_suspending_reason = ""  							
				customer_agreement.save(ignore_permissions=True)


			if payment_history.payment_type == "Normal Payment":
				"""
				deduct assigned bonus and discount from specific agreement
				"""
				if customer_agreement.name == payment_history.assigned_bonus_and_discount:
					customer_agreement.assigned_bonus = float(customer_agreement.assigned_bonus) - float(payment_history.bonus)
					customer_agreement.assigned_discount = float(customer_agreement.assigned_discount) - float(payment_history.discount)

				# if campaign_discount_of_agreements_list and agreement == campaign_discount_of_agreements_list[i].split("/")[0]:
				# 	customer_agreement.discount = campaign_discount_of_agreements_list[i].split("/")[1]
				# 	customer_agreement.assigned_campaign_discount = customer_agreement.assigned_campaign_discount - float(campaign_discount_of_agreements_list[i].split("/")[1])
				# 	customer_agreement.discounted_payments_left = campaign_discount_of_agreements_list[i].split("/")[2]

				if campaign_discount_of_agreements_list and agreement in campaign_discount_of_agreements_dict.keys():
					customer_agreement.discount = float(campaign_discount_of_agreements_dict[agreement][0])
					customer_agreement.assigned_campaign_discount = customer_agreement.assigned_campaign_discount - float(campaign_discount_of_agreements_dict[agreement][0])
					customer_agreement.discounted_payments_left = float(campaign_discount_of_agreements_dict[agreement][1])
				customer_agreement.save(ignore_permissions=True)	
				refund_bonus.append(float(set_values_in_agreement_temporary(agreement,customer.bonus,flag,payments_id_list)))
		#customer.bonus = customer.bonus - sum(refund_bonus) + float(payment_history.bonus)
		customer.bonus = customer.bonus - float(payment_history.new_bonus) + float(payment_history.bonus)
		customer.used_bonus = float(customer.used_bonus) - float(payment_history.bonus)
		customer.refund_to_customer = float(payment_history.cash) + float(payment_history.bank_card) + float(payment_history.bank_transfer) - float(payment_history.bonus) - float(payment_history.discount)

	#customer.receivables = float(payment_history.rental_payment) - float(payment_history.late_fees) - float(payment_history.total_charges)
	#customer.receivables = payment_history.receivables
	customer.receivables = float(customer.receivables) + float(payment_history.receivables) - float(payment_history.receivables_collected)
	customer.old_receivables = payment_history.receivables
	customer.save(ignore_permissions=True)
	
	payment_history.refund = "Yes"
	payment_history.save(ignore_permissions=True)